src/ManageCourses.Api/Data/UserDataHelper.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using GovUk.Education.ManageCourses.Api.Mapping;
+using GovUk.Education.ManageCourses.Domain.DatabaseAccess;
+using GovUk.Education.ManageCourses.Domain.Models;
+
+namespace GovUk.Education.ManageCourses.Api.Data
+{
+    /// <summary>
+    /// This helper class generates the diff between the existing and new dataset.
+    /// Then implements the upsert
+    /// </summary>
+    public class UserDataHelper
+    {
+        private readonly IManageCoursesDbContext _context;
+        private readonly IReadOnlyCollection<McUser> _dataList;
+        private ImportMapper<McUser> _mapper;
+        /// <summary>
+        /// constructor for the class. Initialises the dataset and context.
+        /// Generates the diff (additions, updates, deletions)
+        /// </summary>
+        /// <param name="context">context for the db</param>
+        /// <param name="dataList">new dataset coming from the importer</param>
+        public UserDataHelper(IManageCoursesDbContext context, IReadOnlyCollection<McUser> dataList)
```

Comment:
Change this into a service there is no reason not to.

Reply:
Also inject a logger, so we can have some perception of whats going on.

Reply:
Unit test?

Reply:
All done